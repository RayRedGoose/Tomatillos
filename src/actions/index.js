export const addMovies = movies => ({
  type: 'ADD_MOVIES',
  movies
});

export const addLoaded = isLoaded => ({
  type: 'IS_LOADED',
  isLoaded
});

export const hasError = error => ({
  type: 'HAS_ERROR',
  error
})

export const addRatings = ratings => ({
  type: 'ADD_RATINGS',
  ratings
});

export const addRating = rating => ({
  type: 'ADD_RATING',
  rating
});

export const addUser = user => ({
  type: 'ADD_USER',
  user
});

export const removeUser = () => ({
  type: 'REMOVE_USER',
  user: null
});
