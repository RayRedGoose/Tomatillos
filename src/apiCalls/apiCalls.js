export const getMovies = async () => {
  const url = 'https://rancid-tomatillos.herokuapp.com/api/v1/movies'
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error('There was an error getting movies.');
  }
  const movies = await response.json();
  return movies
}

export const postSignIn = async (user) => {
  const url = 'https://rancid-tomatillos.herokuapp.com/api/v1/login';
  let options = {
    method: 'POST',
    body: JSON.stringify(user),
    headers: {
     'Content-Type': 'application/json'
    }
  };
  const response = await fetch(url, options);
  if (!response.ok) {
    throw new Error('There was a problem signing in')
  }
  const newUser = await response.json();
  console.log(newUer);
  return newUser;
}
