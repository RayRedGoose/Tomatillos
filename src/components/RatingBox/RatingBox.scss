@import "../../variables.scss";

.rating-box {
  h3 {
    font-weight: 400;
    padding: 5px 20px;
    border-bottom: 1px solid $main-accent-color;
    width: fit-content;
    margin: auto;
  }
  div {
    display: flex;
    justify-content: center;
    align-items: center;
    * {
      cursor: pointer;
      margin: 10px 5px;
      &:hover {
        transform: scale(1.2);
      }
      &:active {
        transform: scale(0.8);
      }
    }
    p {
      padding: 0;
      font-size: 24px;
      &:hover {
        color: $titles-color;
        text-shadow: $text-shadow;
      }
    }
    img {
      width: 16px;
      margin-left: 10px;
    }
  }
}
