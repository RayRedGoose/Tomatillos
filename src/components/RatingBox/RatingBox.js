import React, { Component } from 'react';
import './RatingBox.scss';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import cancelIcon from '../../assets/images/cancel.svg';
import { postRating } from '../../apiCalls/apiCalls'
import { addRating } from '../../actions'

export class RatingBox extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isClicked: false
    }
  }

  toggleClicked = () => {
    this.setState({isClicked: !this.state.isClicked})
  }

  chooseRating = async (event) => {
    const {movie_id, addRating} = this.props
    const rating = parseInt(event.target.getAttribute('id').replace('number', ''));
    const ratingFull = {
      movie_id,
      rating
    }
    await postRating(ratingFull, this.props.user_id);
    addRating(ratingFull);
  }

  render() {
    const numbers = new Array(10)
      .fill(undefined)
      .map((el, i) => (
        <p
          key={`number${i+1}`}
          className="rate-number"
          onClick={this.chooseRating}
          id={`number${i+1}`}>{i+1}</p>
      ));

    return (
      (!this.state.isClicked)
        ? <button className="add-rating" onClick={this.toggleClicked}>add rating</button>
        : (
          <section className="rating-box">
            <h3>Choose rating</h3>
            <div>
              {numbers}
              <img onClick={this.toggleClicked} src={cancelIcon} alt="cancel"/>
            </div>
          </section>
        )
    )
  }
}

export const mapDispatchToProps = dispatch => (
  bindActionCreators({
    addRating
  }, dispatch)
)

export default connect(null, mapDispatchToProps)(RatingBox);
