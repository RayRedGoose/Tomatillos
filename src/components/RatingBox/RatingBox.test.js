import React from 'react';
import { shallow } from 'enzyme';
import { RatingBox, mapDispatchToProps } from './RatingBox';
import { postRating } from '../../apiCalls/apiCalls'
import { addRating } from '../../actions'


jest.mock('../../apiCalls/apiCalls');

describe("RatingBox", () => {
  describe("RatingBox component", () => {
    let box, instance;

    beforeEach(() => {

      box = shallow(
        <RatingBox
          user_id={1}
          movie_id={2}
          addRating={jest.fn}/>
      )

      instance = box.instance()
    });

    it("should match snapshot with all data passed in correctly and isClicked state is false", () => {
      expect(box).toMatchSnapshot();
    });

    it("should match snapshot with all data passed in correctly and isClicked state is true", () => {
      box.setState({isClicked: true});
      expect(box).toMatchSnapshot();
    });

    it("should call toggleClicked when button is clicked", () => {
      const spy = jest.spyOn(instance, 'toggleClicked').mockImplementation(() => {});
      instance.forceUpdate();

      box.find('button').simulate('click');

      expect(spy).toHaveBeenCalled();
    });

    it("should call toggleClicked when cancel img is clicked", () => {
      box.setState({isClicked: true});

      const spy = jest.spyOn(instance, 'toggleClicked').mockImplementation(() => {});
      instance.forceUpdate();

      box.find('img').simulate('click');

      expect(spy).toHaveBeenCalled();
    });

    it("should change isClicked state when toggleClicked is called", () => {
      expect(box.state('isClicked')).toEqual(false);

      instance.toggleClicked();

      expect(box.state('isClicked')).toEqual(true);
    });

    describe("chooseRating", () => {
      beforeEach(() => {
        box.setState({isClicked: true});

        postRating.mockImplementation(() => {
          return Promise.resolve({
            id: 1, movie_id: 1, user_id: 1, rating: 6
          })
        })
      });

      it("should call chooseRating when rating number is clicked", () => {
        const spy = jest.spyOn(instance, 'chooseRating').mockImplementation(() => {});
        instance.forceUpdate();

        box.find('#number2').simulate('click');

        expect(spy).toHaveBeenCalled();
      });

      it("should call postRating when chooseRating is called", async () => {
        await instance.chooseRating({
          target: {
            getAttribute: () => 'number1'
          }
        });

        expect(postRating).toHaveBeenCalled();
      });

      it("should call addRating prop when chooseRating is called", async () => {
        const addRating = jest.fn();

        const box = shallow(
          <RatingBox
            user_id={1}
            movie_id={2}
            addRating={addRating}/>
        )

        await box.instance().chooseRating({
          target: {
            getAttribute: () => 'number1'
          }
        });

        expect(addRating).toHaveBeenCalled();
      });
    });
  });

  describe("mapDispatchToProps", () => {
    it('calls dispatch with an addRating action when addRating is called', () => {
      // Setup
      const mockRatingData = {
        id: 1,
        user_id: 1,
        movie_id: 1,
        rating: 7,
        created_at: "2019-12-16",
        updated_at: "2019-12-16"
      };
      const mockDispatch = jest.fn();
      const mappedProps = mapDispatchToProps(mockDispatch);
      const actionToDispatch = addRating(mockRatingData);
      // Execution
      mappedProps.addRating(mockRatingData);
      // Expectaion
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});
