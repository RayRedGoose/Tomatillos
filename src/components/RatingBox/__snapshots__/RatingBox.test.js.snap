// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingBox RatingBox component should match snapshot with all data passed in correctly and isClicked state is false 1`] = `
<button
  className="add-rating"
  onClick={[Function]}
>
  add rating
</button>
`;

exports[`RatingBox RatingBox component should match snapshot with all data passed in correctly and isClicked state is true 1`] = `
<section
  className="rating-box"
>
  <h3>
    Choose rating
  </h3>
  <div>
    <p
      className="rate-number"
      id="number1"
      key="number1"
      onClick={[Function]}
    >
      1
    </p>
    <p
      className="rate-number"
      id="number2"
      key="number2"
      onClick={[Function]}
    >
      2
    </p>
    <p
      className="rate-number"
      id="number3"
      key="number3"
      onClick={[Function]}
    >
      3
    </p>
    <p
      className="rate-number"
      id="number4"
      key="number4"
      onClick={[Function]}
    >
      4
    </p>
    <p
      className="rate-number"
      id="number5"
      key="number5"
      onClick={[Function]}
    >
      5
    </p>
    <p
      className="rate-number"
      id="number6"
      key="number6"
      onClick={[Function]}
    >
      6
    </p>
    <p
      className="rate-number"
      id="number7"
      key="number7"
      onClick={[Function]}
    >
      7
    </p>
    <p
      className="rate-number"
      id="number8"
      key="number8"
      onClick={[Function]}
    >
      8
    </p>
    <p
      className="rate-number"
      id="number9"
      key="number9"
      onClick={[Function]}
    >
      9
    </p>
    <p
      className="rate-number"
      id="number10"
      key="number10"
      onClick={[Function]}
    >
      10
    </p>
    <img
      alt="cancel"
      onClick={[Function]}
      src="cancel.svg"
    />
  </div>
</section>
`;
